naming/program/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{msg, program_error::ProgramError, pubkey::Pubkey};
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = Pubkey;
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum IssuerKind {
+    Metadata,       // Metadata authority has sole control
+    Issuer(Pubkey), // Issuing authority has sole control
+    Both(Pubkey),   // Issuing and metadata authority must sign
+    Either(Pubkey), // Issuing or metadata authority must sign
+    Open,           // First come, first served. Terminal state
+    Frozen,         // No further issuance allowed.  Terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub enum OwnerKind {
+    Account(Pubkey),
+    Name(String), // Usefull for easy ownership transfer of a whole sub-tree of domain names
+    Frozen,       // Metadata is locked, terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub struct NameRegistryState {
+    pub status: NameStatus,
+    // This is the complete name, including all parents
+    pub name: String,
+    pub metadata_authority: OwnerKind,
+    pub issuing_authority: IssuerKind,
+    pub values: Vec<(ValueType, String)>, // TODO use hashmaps?
```

Comment:
I think we want one "Value" per account, like my design sketch at https://github.com/solana-labs/solana-program-library/pull/1600#issuecomment-820849453

The problem with multiple values is that it makes searching for data over RPC very inefficient as the `getProgramAccounts` [memcmp](https://docs.solana.com/developing/clients/jsonrpc-api#filters) filter can no longer be used because position of the `ValueType` or value data of interest is no longer fixed in the account data.

The value data itself should just be bytes at this level as well.  Depending on the `ValueType`, the data may be a String but not necessarily.

Reply:
I agree that the values can be stored as bytes directly, I'm going to change it now. I would be happy to discuss the rest with you directly :) I think that this is a core decision we have to take from which the other questions can be at least partially answered.